man/vnstatd.1
.TH VNSTATD 1 "JULY 2019" "version 2.3" "User Manuals"
.SH NAME
vnstatd \- daemon based database updating for vnStat

.SH SYNOPSIS

.B vnstatd
.RB [ \-Ddnpsv? ]
.RB [ \-\-alwaysadd ]
.RB [ \-\-config
.IR file ]
.RB [ \-\-daemon ]
.RB [ \-\-debug ]
.RB [ \-g
.IR group ]
.RB [ \-\-group
.IR group ]
.RB [ \-\-help ]
.RB [ \-\-noadd ]
.RB [ \-\-nodaemon ]
.RB [ \-\-pidfile
.IR file ]
.RB [ \-\-sync ]
.RB [ \-\-u
.IR user ]
.RB [ \-\-user
.IR user ]
.RB [ \-\-version ]

.SH DESCRIPTION

The purpose of
.B vnstatd
is to provide a flexible and robust way for updating the database that
.BR vnstat (1)
uses. The availability of each interface is automatically tracked which
removes the need for additional scripts to be implemented and called when
an interface comes online or goes offline.
.PP
.B vnstatd
is the command for starting the daemon. The daemon can either fork
itself to run as a background process or stay attached to the terminal.
It supports logging to a user selectable file or using syslog.
.PP
Once started, the daemon will read
.BR vnstat.conf (5)
if available and then check if there is a database present
in the database directory that has been specified in the configuration
file. By default, if no database is found, a database will be created
during startup with entries for all available interfaces excluding pseudo
interfaces lo, lo0 and sit0. This automatic database entry creation behaviour
can be disabled using the
.B --noadd
option. Alternatively, it is possible to allow the daemon to create new
database entries whenever previously unseen interfaces become visible using the
.B --alwaysadd
option.
.PP
The daemon will proceed to track the availability of monitored interfaces,
process the interface traffic statistics and write new values to the database
at a configured interval. As a result, the daemon ends up spending most
of the time sleeping between updates.

.SH OPTIONS

.TP
.B "--alwaysadd"
Enable automatic creation of new database entries for previously unseen interfaces
even if the database directory already contains a database when the daemon
is started. New database entries will also get created for new interfaces seen while
the daemon is running. Pseudo interfaces lo, lo0 and sit0 are excluded from
getting added.

.TP
.BI "--config " file
Use
.I file
as configuration file instead of using automatic configuration file search
functionality.

.TP
.B "-d, --daemon"
Fork process to background and run as a daemon.

.TP
.B "-D, --debug"
Provide additional output for debug purposes. The process will stay
attached to the terminal for output.

.TP
.BI "-g, --group " group
Set daemon process group to
.I group
during startup.
.I group
can be either the name of the group or a numerical group id. This option
can only be used when the process is started as root.

.TP
.B "--noadd"
Disable automatic creation of new database entries for all available interfaces
if the daemon is started with no database found. Pseudo interfaces
lo, lo0 and sit0 are excluded from getting added.

.TP
.B "-n, --nodaemon"
Stay in foreground attached to the current terminal and start the update
process.

.TP
.BI "-p, --pidfile " file
Write the process id to
.I file
and use it for locking so that another instance of the daemon cannot
be started if the same
.I file
is specified.

.TP
.B "-s, --sync"
Synchronize internal counters in the database with interface
counters for all available interfaces before starting traffic monitoring.
Use this option if the traffic between the previous shutdown
and the current startup of the daemon needs to be ignored. This option
isn't required in normal use because the daemon will automatically synchronize
the internal counters after a system reboot, if enough time has passed
since the daemon was previously running or if the internal counters are
clearly out of sync.

.TP
.BI "-u, --user " user
Set daemon process user to
.I user
during startup.
.I user
can be either the login of the user or a numerical user id. This option
can only be used when the process is started as root.

.TP
.B "-v, --version"
Show current version of the daemon executable.

.TP
.B "-?, --help"
Show a command option summary.

.SH CONFIGURATION

The behaviour of the daemon is configured mainly using the configuration
keywords
.B "UpdateInterval, PollInterval"
and
.B SaveInterval
in the configuration file.

.PP
.B UpdateInterval
defines in seconds how often the interface data is fetched and updated.
This is similar to the run interval for alternative cron based updating.
However, the difference is that the data doesn't directly get written to disk
during updates.

.PP
.B PollInterval
defines in seconds how often the list of available interfaces is checked
for possible changes. The minimum value is 2 seconds and the maximum 60
seconds.
.B PollInterval
also defines the resolution for other intervals.

.PP
.B SaveInterval
defines in minutes how often cached interface data is written to disk.
A write can only occur during the updating of interface data. Therefore,
the value should be a multiple of
.B UpdateInterval
with a maximum value of 60 minutes.

.PP
The default values of
.B UpdateInterval
30,
.B SaveInterval
5 and
.B PollInterval
5 are usually suitable for most systems and provide a similar behaviour
as cron based updating does but with a better resolution for interface
changes and fast interfaces.

.PP
For embedded and/or low power systems more tuned configurations are possible.
In such cases if the interfaces are mostly static the
.B PollInterval
can be increased to around 10-30 seconds and
.B UpdateInterval
set to 60 seconds. Higher values up to 300 seconds are possible if the
interface speed is 10 Mbit or less.
.B SaveInterval
can be increased for example to 15, 30 or even 60 minutes depending on how
often the data needs to be viewed.

.SH SIGNALS

The daemon is listening to signals
.B "SIGHUP, SIGINT"
and
.B SIGTERM.
Sending the
.B SIGHUP
signal to the daemon will cause cached data to be written to disk,
a rescan of the database directory and a reload of settings from the
configuration file. However, the pid file location will not be changed
even if it's configuration setting has been modified.

.PP
.B SIGTERM
and
.B SIGINT
signals will cause the daemon to write all cached data to disk and
then exit.

.SH FILES

.TP
.I /var/lib/vnstat/
Default database directory.

.TP
.I /etc/vnstat.conf
Config file that will be used unless
.I $HOME/.vnstatrc
exists. See the configuration chapter and
.BR vnstat.conf (5)
for more information.

.TP
.I /var/log/vnstat.log
Log file that will be used if logging to file is enable and no other file
is specified in the config file.

.TP
.I /var/run/vnstat.pid
File used for storing the process id if no other file is specified in the
configuration file or using the command line parameter.

.SH RESTRICTIONS

Updates need to be executed at least as often as it is possible for the interface
to generate enough traffic to overflow the kernel interface traffic counter. Otherwise,
it is possible that some traffic won't be seen. With 32-bit kernels, the maximum time
between two updates depends on how fast the interface can transfer 4 GiB. Calculated
theoretical times are:
.RS
.TS
l l.
10 Mbit:        54 minutes
100 Mbit:        5 minutes
1000 Mbit:      30 seconds
.TE
.RE

However, for 1000 Mbit interfaces updating once every minute is usually a
usable solution if a shorter update interval can't be used.
.PP
Virtual and aliased interfaces cannot be monitored because the kernel doesn't
provide traffic information for that type of interfaces. Such interfaces are
usually named eth0:0, eth0:1, eth0:2 etc. where eth0 is the actual interface
being aliased.

.SH AUTHOR

Teemu Toivola <tst at iki dot fi>

.SH "SEE ALSO"

.BR vnstat (1),
.BR vnstati (1),
.BR vnstat.conf (5),
.BR signal (7)


CHANGES
2.3 / 10-Jul-2019

 - Fixed
   - Traffic rate wasn't being shown for the last entry of most lists during
     the first update of the next period (first SaveInterval minutes of each
     new hour, day, month and year)
   - Systemd example service file could result in database file write issues
     if the used systemd version supported ProtectSystem=strict but didn't
     support StateDirectory (issue seen at least with systemd 232 in Debian 9)
   - Debian and Red Hat init.d example files had wrong path for the pid file
   - Interfaces could end up staying marked as 'disabled' in the database even
     after becoming back active and monitored, only the shown status was wrong
     without resulting in any data loss
   - Some image outputs didn't allocate enough height for the image resulting
     in the last row having less space below it than intended
 - New
   - Automatic interface selection when the Interface configuration setting
     is left empty (new default)
   - Add configuration option DatabaseWriteAheadLogging to enable SQLite
     Write-Ahead Logging mode which may provide some disk i/o benefits,
     see https://www.sqlite.org/wal.html for more details and note that
     SQLite 3.22.0 or later is required to support read-only operations
   - Add configuration option DatabaseSynchronous for changing the SQLite
     setting of the "synchronous" flag, see
     https://www.sqlite.org/pragma.html#pragma_synchronous for more details
   - Show warning in log if writing cached data to database is slow
   - Try database query for up to 5 seconds when database is busy or locked
     instead of giving up immediately
   - Continue daemon process execution with data caching if database writes
     fail due to disk being full


2.2 / 28-Apr-2019

 - Fixed
   - O_CLOEXEC undeclared error when compiling with glibc older than 2.12
   - Image output --headertext wasn't being applied to some output types
   - DefaultDecimals setting wasn't controlling all outputs
   - Traffic rate for the last entry of the 5 minute resolution list wasn't
     being correctly calculated
 - New
   - Detection of 32-bit / 64-bit kernel interface counters for improving
     traffic calculations in rollover situations, cfg: 64bitInterfaceCounters
   - Possibility in queries to merge data from multiple interfaces to one
     output with interface1+interface2 syntax
   - Highlight current day in top list when visible
   - Changes in interface specific detected bandwidth limits are reported
     in the log
   - UpdateInterval value is automatically adjusted during daemon startup if
     found to be too high compared to the value of MaxBandwidth
 - Notes
   - Added clarification to the man page that the 'jsonversion' and
     'xmlversion' fields in the --json and --xml outputs should be seen as
     the API version and compared to the previously used version in 3rd party
     implementations for verifying compatibility
   - Default value of UpdateInterval is now 20 instead of previous 30 seconds


2.1 / 6-Dec-2018

 - Fixed
   - --oneline could segmentation fault if database had total traffic values
     set but didn't have any daily or monthly traffic (Arch bug #60588)
   - No proper error message was produced when querying interface which didn't
     exist in the database
   - Several typos and spelling mistakes in documentation
   - Daemon:waittimesync_does_not_wait_with_new_interfaces test was failing if
     the system used a time zone with a negative numeric value
   - Traffic estimate and average rate for ongoing month was not being
     calculated correctly when MonthRotate had a value greater than 1
   - Accuracy of traffic estimate and average rate for the ongoing month could
     be slightly skewed depending on the used time zone
 - New
   - Remove possible PollInterval and UpdateInterval drift over time
   - Sync SaveInterval to full minutes


2.0 / 21-Oct-2018

 - Fixed
   - Data rate calculations when using SI decimal prefixes
   - Workaround for filled arcs not drawn correctly with some versions
     of libgd (libgd/libgd#351)
 - New
   - C struct dump database has been replaced with sqlite database,
     full data import included from legacy databases of versions 1.3 - 1.18
   - User configurable data retention durations for all resolutions with
     following changed defaults:
                   v1.18     v2.0
       5 minutes     n/a      576   (48 hours)
       hours          24       96   (4 days)
       days           30       62   (2 full months)
       months         12       25   (2 full years)
       years         n/a      all
       top days       10       20
   - Option to select units to use SI decimal prefixes using calculations
     with powers of 1000 (IEC and JEDEC prefixes use powers of 1024)
   - Yearly and 5 minute resolution statistics
   - Pebibyte and exbibyte support
   - Options the configure the length of each list output
   - Option to limit the length of --json and --xml outputs
   - Hourly list output replaces the hourly graph when the -h parameter is used
   - Hourly graph is accessible with -hg / --hoursgraph parameter
   - Options -b / --begin and -e / --end for specifying query time range
   - Configuration comes with ISO 8601 date formats as new defaults
 - Removed
   - Weekly output
   - Merge of data from multiple interfaces to one output
   - --testkernel
   - --exportdb, use --oneline, --json or --xml instead
   - --importdb, -u / --update and related options using vnstat command
   - Old style (default in versions up to 1.7) summary layout in image output


1.18 / 4-Mar-2018

 - Fix: Misspellings, compiler warnings and possible error situations
   (pull requests by Christian Göttsche)
 - Hardening added to examples/systemd/vnstat.service
   (pull request by Christian Göttsche)
 - Add RTC sanity check and sync wait to daemon startup, the possible wait
   duration can be configured using the TimeSyncWait keyword

src/vnstatd.c
/*
vnStat daemon - Copyright (c) 2008-2019 Teemu Toivola <tst@iki.fi>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2 dated June, 1991.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#include "common.h"
#include "datacache.h"
#include "dbsql.h"
#include "cfg.h"
#include "ibw.h"
#include "id.h"
#include "daemon.h"
#include "vnstatd.h"

int main(int argc, char *argv[])
{
	int currentarg;
	uint32_t previflisthash;
	uint64_t temp;
	DSTATE s;

	initdstate(&s);

	/* early check for debug and config parameter */
	if (argc > 1) {
		for (currentarg = 1; currentarg < argc; currentarg++) {
			if ((strcmp(argv[currentarg], "-D") == 0) || (strcmp(argv[currentarg], "--debug") == 0)) {
				debug = 1;
				printf("Debug enabled, vnstatd %s\n", VERSION);
			} else if (strcmp(argv[currentarg], "--config") == 0) {
				if (currentarg + 1 < argc) {
					strncpy_nt(s.cfgfile, argv[currentarg + 1], 512);
					if (debug)
						printf("Used config file: %s\n", s.cfgfile);
					currentarg++;
					continue;
				} else {
					printf("Error: File for --config missing.\n");
					return 1;
				}
			}
		}
	}

	timeused_debug("daemon_startup", 1);

	/* load config if available */
	if (!loadcfg(s.cfgfile)) {
		return 1;
	}
	if (!ibwloadcfg(s.cfgfile)) {
		return 1;
	}

	/* init config settings */
	strncpy_nt(s.user, cfg.daemonuser, 33);
	strncpy_nt(s.group, cfg.daemongroup, 33);
	s.updateinterval = cfg.updateinterval;
	s.saveinterval = cfg.saveinterval * 60;

	/* parse parameters, maybe not the best way but... */
	for (currentarg = 1; currentarg < argc; currentarg++) {
		if (debug)
			printf("arg %d: \"%s\"\n", currentarg, argv[currentarg]);
		if ((strcmp(argv[currentarg], "-?") == 0) || (strcmp(argv[currentarg], "--help") == 0)) {
			break;
		} else if (strcmp(argv[currentarg], "--config") == 0) {
			/* config has already been parsed earlier so nothing to do here */
			currentarg++;
			continue;
		} else if ((strcmp(argv[currentarg], "-D") == 0) || (strcmp(argv[currentarg], "--debug") == 0)) {
			debug = 1;
		} else if ((strcmp(argv[currentarg], "-d") == 0) || (strcmp(argv[currentarg], "--daemon") == 0)) {
			s.rundaemon = 1;
			s.showhelp = 0;
		} else if ((strcmp(argv[currentarg], "-n") == 0) || (strcmp(argv[currentarg], "--nodaemon") == 0)) {
			s.showhelp = 0;
		} else if ((strcmp(argv[currentarg], "-s") == 0) || (strcmp(argv[currentarg], "--sync") == 0)) {
			s.sync = 1;
		} else if ((strcmp(argv[currentarg], "-u") == 0) || (strcmp(argv[currentarg], "--user") == 0)) {
			if (currentarg + 1 < argc) {
				strncpy_nt(s.user, argv[currentarg + 1], 33);
				if (debug)
					printf("Requested user: \"%s\"\n", s.user);
				currentarg++;
				continue;
			} else {
				printf("Error: User for --user missing.\n");
				return 1;
			}
		} else if ((strcmp(argv[currentarg], "-g") == 0) || (strcmp(argv[currentarg], "--group") == 0)) {
			if (currentarg + 1 < argc) {
				strncpy_nt(s.group, argv[currentarg + 1], 33);
				if (debug)
					printf("Requested group: \"%s\"\n", s.group);
				currentarg++;
				continue;
			} else {
				printf("Error: Group for --group missing.\n");
				return 1;
			}
		} else if (strcmp(argv[currentarg], "--noadd") == 0) {
			s.noadd = 1;
		} else if (strcmp(argv[currentarg], "--alwaysadd") == 0) {
			s.alwaysadd = 1;
		} else if ((strcmp(argv[currentarg], "-v") == 0) || (strcmp(argv[currentarg], "--version") == 0)) {
			printf("vnStat daemon %s by Teemu Toivola <tst at iki dot fi>\n", getversion());
			return 0;
		} else if ((strcmp(argv[currentarg], "-p") == 0) || (strcmp(argv[currentarg], "--pidfile") == 0)) {
			if (currentarg + 1 < argc) {
				strncpy_nt(cfg.pidfile, argv[currentarg + 1], 512);
				cfg.pidfile[511] = '\0';
				if (debug)
					printf("Used pid file: %s\n", cfg.pidfile);
				currentarg++;
				continue;
			} else {
				printf("Error: File for --pidfile missing.\n");
				return 1;
			}
		} else {
			printf("Unknown arg \"%s\". Use --help for help.\n", argv[currentarg]);
			return 1;
		}
	}

	if (s.noadd && s.alwaysadd) {
		printf("Error: --noadd and --alwaysadd can't both be used at the same time.\n");
		return 1;
	}

	if (s.rundaemon && debug) {
		printf("Error: --daemon and --debug can't both be used at the same time.\n");
		return 1;
	}

	/* show help if nothing else was asked to be done */
	if (s.showhelp) {
		showhelp();
		return 0;
	}

	preparedirs(&s);

	/* set user and/or group if requested */
	setgroup(s.group);
	setuser(s.user);

	if (!db_open_rw(1)) {
		printf("Error: Failed to open database \"%s/%s\" in read/write mode.\n", cfg.dbdir, DATABASEFILE);
		printf("Exiting...\n");
		exit(EXIT_FAILURE);
	}

	detectboot(&s);
	preparedatabases(&s);

	if (!db_removeoldentries()) {
		printf("Error: Database \"%s/%s\" cleanup failed: %s\n", cfg.dbdir, DATABASEFILE, strerror(errno));
		printf("Exiting...\n");
		exit(EXIT_FAILURE);
	}

	setsignaltraps();

	/* start as daemon if requested, debug can't be enabled at the same time */
	if (s.rundaemon && !debug) {
		if (!db_close()) {
			printf("Error: Failed to close database \"%s/%s\" before starting daemon: %s\n", cfg.dbdir, DATABASEFILE, strerror(errno));
			printf("Exiting...\n");
			exit(EXIT_FAILURE);
		}
		noexit++;
		daemonize();
		if (!db_open_rw(0)) {
			snprintf(errorstring, 1024, "Failed to reopen database \"%s/%s\": %s", cfg.dbdir, DATABASEFILE, strerror(errno));
			printe(PT_Error);
			exit(EXIT_FAILURE);
		}
	}

	timeused_debug("daemon_startup", 0);
	s.running = 1;

#if defined(__linux__)
#if HAVE_DECL_IFLA_STATS64
	snprintf(errorstring, 1024, "vnStat daemon %s started. (pid:%d uid:%d gid:%d 64-bit)", getversion(), (int)getpid(), (int)getuid(), (int)getgid());
#else
	snprintf(errorstring, 1024, "vnStat daemon %s started. (pid:%d uid:%d gid:%d 32-bit)", getversion(), (int)getpid(), (int)getuid(), (int)getgid());
#endif
#else
	snprintf(errorstring, 1024, "vnStat daemon %s started. (pid:%d uid:%d gid:%d)", getversion(), (int)getpid(), (int)getuid(), (int)getgid());
#endif
	printe(PT_Info);

#if !HAVE_DECL_SQLITE_CHECKPOINT_RESTART
	if (cfg.waldb) {
		snprintf(errorstring, 1024, "DatabaseWriteAheadLogging is enabled but used libsqlite3 does not support it");
		printe(PT_Warning);
	}
#endif

	/* warmup */
	if (s.dbifcount == 0) {
		filldatabaselist(&s);
		s.prevdbsave = 0;
	}
	while (s.running && s.dbifcount && waittimesync(&s)) {
		if (intsignal) {
			handleintsignals(&s);
		} else {
			sleep(5);
		}
	}

	/* main loop */
	while (s.running) {

		s.current = time(NULL);

		/* track interface status only if at least one database exists */
		if (s.dbifcount != 0) {
			previflisthash = s.iflisthash;
			interfacechangecheck(&s);
			if (s.alwaysadd && s.iflisthash != previflisthash && previflisthash != 0) {
				temp = s.dbifcount;
				s.dbifcount += addinterfaces(&s);
				if (temp != s.dbifcount) {
					datacache_status(&s.dcache);
				}
			}
		}

		/* do update only if enough time has passed since the previous update */
		if ((s.current - s.prevdbupdate) >= s.updateinterval) {

			s.updateinterval = cfg.updateinterval;

			if (debug) {
				debugtimestamp();
				datacache_debug(&s.dcache);
				ibwlist();
			}

			/* fill database list if cache is empty */
			if (s.dbifcount == 0) {
				filldatabaselist(&s);

				/* update data cache */
			} else {
				s.prevdbupdate = s.current - (s.current % s.updateinterval);

				adjustsaveinterval(&s);
				checkdbsaveneed(&s);

				processdatacache(&s);

#if HAVE_DECL_SQLITE_CHECKPOINT_RESTART
				if (cfg.waldb && (s.current - s.prevwaldbcheckpoint) >= WALDBCHECKPOINTINTERVALMINS * 60) {
					db_walcheckpoint();
					s.prevwaldbcheckpoint = s.current;
				}
#endif

				if (debug) {
					printf("\n");
				}
			}
		}

		if (s.running && intsignal == 0) {
			sleep((unsigned int)(cfg.pollinterval - (time(NULL) % cfg.pollinterval)));
		}

		if (intsignal) {
			handleintsignals(&s);
		}
	}

	flushcachetodisk(&s);
	db_close();

	datacache_clear(&s.dcache);
	ibwflush();

	if (s.rundaemon && !debug) {
		close(pidfile);
		unlink(cfg.pidfile);
	}

	return 0;
}

void showhelp(void)
{
	printf("vnStat daemon %s by Teemu Toivola <tst at iki dot fi>\n\n", getversion());

	printf("      -d, --daemon             fork process to background\n");
	printf("      -n, --nodaemon           stay in foreground attached to the terminal\n\n");

	printf("      -s, --sync               sync interface counters on first update\n");
	printf("      -D, --debug              show additional debug and disable daemon\n");
	printf("      -?, --help               show this help\n");
	printf("      -v, --version            show version\n");
	printf("      -p, --pidfile <file>     select used pid file\n");
	printf("      -u, --user <user>        set daemon process user\n");
	printf("      -g, --group <group>      set daemon process group\n");
	printf("      --config <config file>   select used config file\n");
	printf("      --noadd                  don't add found interfaces if no dbs are found\n");
	printf("      --alwaysadd              always add new interfaces even when some dbs exist\n\n");

	printf("See also \"man vnstatd\".\n");
}
